src/python/pants/backend/python/dependency_inference/import_parser.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import ast as ast3
+import re
+import sys
+import warnings
+from dataclasses import dataclass
+from typing import Set
+
+from typed_ast import ast27
+
+from pants.util.ordered_set import FrozenOrderedSet
+from pants.util.strutil import ensure_text
+
+
+class ImportParseError(ValueError):
+    pass
+
+
+@dataclass(frozen=True)
+class ParsedPythonImports:
+    """All the discovered imports from a Python source file.
+
+    Explicit imports are imports from `import x` and `from module import x` statements. Inferred
+    imports come from strings that look like module names, such as
+    `importlib.import_module("example.subdir.Foo")`.
+    """
+
+    explicit_imports: FrozenOrderedSet[str]
+    inferred_imports: FrozenOrderedSet[str]
+
+
+def parse_file(source_code: str, *, module_name: str):
+    try:
+        # NB: The Python 3 ast is generally backwards-compatible with earlier versions. The only
+        # breaking change is `async` `await` becoming reserved keywords in Python 3.7 (deprecated
+        # in 3.6). If the std-lib fails to parse, we could use typed-ast to try parsing with a
+        # target version of Python 3.5, but we don't because Python 3.5 is almost EOL and has very
+        # low usage.
+        # We will also fail to parse Python 3.8 syntax if Pants is run with Python 3.6 or 3.7.
+        # There is no known workaround for this, beyond users changing their `./pants` script to
+        # always use >= 3.8.
+        tree = ast3.parse(source_code)
+        visitor_cls = _Py3AstVisitor if sys.version_info[:2] < (3, 8) else _Py38AstVisitor
+        return tree, visitor_cls
+    except Exception as e:
+        try:
+            return ast27.parse(source_code), _Py27AstVisitor
+        except Exception:
+            raise ImportParseError(f"Failed to parse source code for {module_name}:\n{e!r}")
+
+
+def find_python_imports(source_code: str, *, module_name: str) -> ParsedPythonImports:
+    tree, ast_visitor_cls = parse_file(source_code, module_name=module_name)
+    ast_visitor = ast_visitor_cls(module_name)
+    with warnings.catch_warnings():
+        # We often encounter this deprecation warning when parsing files. It's noisy for us to
+        # display.
```

Comment:
catch_warnings isn't threadsafe: https://docs.python.org/3/library/warnings.html#warnings.catch_warnings

If we'd like to ignore this more generally, we should do so globally, once.

Reply:
I'll remove it. It was copied over from Toolchain without much thought.